iridauploader/core/upload.py

```diff
@@ -0,0 +1,253 @@
+"""
+This file contains the single entry functions to upload data that are called via the command line and the gui
+These functions return exit codes that get passed up the chain to whatever function calls the uploader (i.e. bash)
+
+It also contains the functions for starting/stopping logging to a directory
+"""
+
+import logging
+
+from iridauploader import VERSION_NUMBER
+import iridauploader.api as api
+import iridauploader.config as config
+import iridauploader.parsers as parsers
+import iridauploader.progress as progress
+from iridauploader.model import DirectoryStatus
+
+from . import api_handler, parsing_handler, logger, exit_return, upload_helpers
+
+
+def upload_run_single_entry(directory, force_upload=False, upload_mode=None):
+    """
+    This function acts as a single point of entry for uploading a directory
+
+    Handles getting a directories run status, and running if conditions are met (valid run, new run or forced upload).
+
+    :param directory: Directory of the sequencing run to upload
+    :param force_upload: When set to true, the upload status file will be ignored and file will attempt to be uploaded
+    :param upload_mode: String with upload mode to use. When None, default is used.
+    :return: ExitReturn
+    """
+
+    directory_status = parsing_handler.get_run_status(directory)
+    delay_minutes = config.read_config_option("delay", expected_type=int)
+
+    # Check that directory is writeable, or readonly mode is enabled
+    if upload_helpers.directory_has_readonly_conflict(directory_status.directory):
+        error_msg = 'Directory cannot be written to. Please check permissions or use readonly mode'
+        logging.error(error_msg)
+        return exit_error(error_msg)
+    # Check if a run is invalid, an invalid run cannot be uploaded.
+    elif directory_status.status_equals(DirectoryStatus.INVALID):
+        error_msg = "ERROR! Run in directory {} is invalid. Returned with message: '{}'".format(
+            directory_status.directory, directory_status.message)
+        logging.error(error_msg)
+        return exit_error(error_msg)
+    # Check if run is new, if delay config is > 0, delay the run, exit with success
+    elif directory_status.status_equals(DirectoryStatus.NEW):
+        if delay_minutes > 0:
+            upload_helpers.set_run_delayed(directory_status)
+            logging.info("Run has been delayed for {} minutes.".format(delay_minutes))
+            return exit_success()
+    # If run was delayed, check if run can now be uploaded, if not, exit with success
+    elif directory_status.status_equals(DirectoryStatus.DELAYED):
```

Comment:
I feel like this big conditional has made this method and `batch_upload_single_entry` overly messy.  They used to be pretty small tight methods for the tasks they're doing.  I think this should be refactored somehow,

Does it make sense to take that code out into another component or something?  Maybe move it out into the `upload_helpers` module or something so this method can just ask "is this run ready to go?"